src/lib/Translation/UserLanguagePreferenceProvider.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace EzSystems\EzPlatformAdminUi\Translation;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Provides list of user-preferred languages.
+ */
+class UserLanguagePreferenceProvider
+{
+    /** @var \Symfony\Component\HttpFoundation\RequestStack */
+    private $requestStack;
+
+    /** @var array */
+    private $languageCodesMap;
+
+    public function __construct(RequestStack $requestStack, array $languageCodesMap)
+    {
+        $this->requestStack = $requestStack;
+        $this->languageCodesMap = $languageCodesMap;
+    }
+
+    /**
+     * Return a list of user's browser preferred locales directly from Accept-Language header.
+     *
+     * @param \Symfony\Component\HttpFoundation\Request request to retrieve information from, use current if null
+     *
+     * @return string[]
+     */
+    public function getPreferredLocales(Request $request = null): array
+    {
+        $request = $request ?? $this->requestStack->getCurrentRequest();
+        $preferredLocales = $request->headers->get('Accept-Language') ?? '';
+        $preferredLocales = array_reduce(
+            explode(',', $preferredLocales),
+            function ($res, $el) {
+                [$l, $q] = array_merge(explode(';q=', $el), [1]);
+                $res[$l] = (float) $q;
+
```

Comment:
I know `$l` and `$q` are temporary local variables, but to avoid questions like: "Is this PERL?" or "Do we minify PHP files by hand already?", it would be good to make variable names a bit more descriptive.

Reply:
Has been improved in ezsystems/ezpublish-kernel#2507